crates/lsp/src/tower_lsp.rs

```diff
@@ -65,31 +66,97 @@ pub(crate) enum LspRequest {
 }
 
 #[allow(clippy::large_enum_variant)]
-#[derive(Debug)]
+#[derive(Debug, IntoStaticStr)]
 pub(crate) enum LspResponse {
     Initialize(InitializeResult),
     DocumentFormatting(Option<Vec<TextEdit>>),
     Shutdown(()),
     AirViewFile(String),
 }
 
+impl std::fmt::Display for LspNotification {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.write_str(self.into())
+    }
+}
+impl std::fmt::Display for LspRequest {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.write_str(self.into())
+    }
+}
+impl std::fmt::Display for LspResponse {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.write_str(self.into())
+    }
+}
+
+impl LspNotification {
+    fn trace(&self) -> TraceLspNotification {
+        TraceLspNotification { inner: self }
+    }
+}
+impl LspRequest {
+    fn trace(&self) -> TraceLspRequest {
+        TraceLspRequest { inner: self }
+    }
+}
+impl LspResponse {
+    fn trace(&self) -> TraceLspResponse {
+        TraceLspResponse { inner: self }
+    }
+}
+
+struct TraceLspNotification<'a> {
+    inner: &'a LspNotification,
+}
+struct TraceLspRequest<'a> {
+    inner: &'a LspRequest,
+}
+struct TraceLspResponse<'a> {
+    inner: &'a LspResponse,
+}
+
+impl std::fmt::Debug for TraceLspNotification<'_> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self.inner {
+            LspNotification::DidOpenTextDocument(params) => {
+                // Ignore the document itself in trace logs
+                f.debug_tuple(self.inner.into())
+                    .field(&params.text_document.uri)
+                    .field(&params.text_document.version)
+                    .field(&params.text_document.language_id)
+                    .finish()
+            }
+            _ => std::fmt::Debug::fmt(self.inner, f),
+        }
+    }
+}
+
+impl std::fmt::Debug for TraceLspRequest<'_> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        std::fmt::Debug::fmt(self.inner, f)
+    }
+}
+
+impl std::fmt::Debug for TraceLspResponse<'_> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        std::fmt::Debug::fmt(self.inner, f)
+    }
+}
+
 #[derive(Debug)]
 struct Backend {
     /// Channel for communication with the main loop.
     events_tx: TokioUnboundedSender<Event>,
 
-    /// Channel for communication with the auxiliary loop.
-    /// Used for latency sensitive logging.
-    auxiliary_event_tx: AuxiliaryEventSender,
-
     /// Handle to main loop. Drop it to cancel the loop, all associated tasks,
     /// and drop all owned state.
     _main_loop: tokio::task::JoinSet<()>,
 }
 
 impl Backend {
     async fn request(&self, request: LspRequest) -> anyhow::Result<LspResponse> {
-        self.log_info(format!("Incoming: {request:#?}"));
+        tracing::trace!("Incoming:\n{request:#?}", request = request.trace());
```

Comment:
I think we also discussed a `tracing::info()` message with the message type so we can better make sense of user logs before they enable trace logging?